src/client/Microsoft.Identity.Client/Platforms/Android/AndroidBrokerHelper.cs

```diff
@@ -173,29 +172,44 @@ public void GetBrokerAccountInfo(IDictionary<string, string> brokerPayload, Acti
 
             if (result == null)
             {
-                _logger.Info("Android account manager AddAccount didn't return any accounts. ");
-                throw new MsalException(MsalError.NoAndroidBrokerAccountFound, "Please add the selected account to the broker");
+                _logger.Info("Android account manager didn't return any accounts.");
+                throw new MsalUiRequiredException(MsalError.NoAndroidBrokerAccountFound, MsalErrorMessage.NoAndroidBrokerAccountFound);
             }
 
             Bundle bundleResult = (Bundle)result?.Result;
             var accounts = bundleResult?.GetString(BrokerConstants.BrokerAccounts);
 
+            string username = GetValueFromBrokerPayload(brokerPayload, BrokerParameter.Username);
+            string homeAccountId = GetValueFromBrokerPayload(brokerPayload, BrokerParameter.HomeAccountId);
+            string localAccountId = GetValueFromBrokerPayload(brokerPayload, BrokerParameter.LocalAccountId);
+
             if (!string.IsNullOrEmpty(accounts))
             {
                 dynamic authResult = JArray.Parse(accounts);
 
                 foreach (JObject account in authResult)
                 {
                     var accountData = account[BrokerResponseConst.Account];
-                    if ((accountData[BrokerResponseConst.UserName]).ToString() == loginHint)
+
+                    if ((accountData[BrokerResponseConst.UserName]).ToString() == username)
+                    {
+                        brokerPayload[BrokerParameter.HomeAccountId] = accountData[BrokerResponseConst.HomeAccountId].ToString();
+                        brokerPayload[BrokerParameter.LocalAccountId] = accountData[BrokerResponseConst.LocalAccountId].ToString();
+                        _logger.Info("Found broker account in Android account manager.");
+                        return;
+                    }
+                    
+                    if ((accountData[BrokerResponseConst.HomeAccountId]).ToString() == homeAccountId &&
+                         (accountData[BrokerResponseConst.LocalAccountId]).ToString() == localAccountId)
                     {
-                        brokerPayload.Add(BrokerParameter.HomeAccountId, accountData[BrokerResponseConst.HomeAccountId].ToString());
-                        brokerPayload.Add(BrokerParameter.LocalAccountId, accountData[BrokerResponseConst.LocalAccountId].ToString());
-                        var acc = account;
+                        _logger.Info("Found broker account in Android account manager.");
                         return;
                     }
                 }
             }
+
+            _logger.Info("The requested account does not exist in the Android account manager.");
+            throw new MsalUiRequiredException(MsalError.NoAndroidBrokerAccountFound, MsalErrorMessage.NoAndroidBrokerAccountFound);
```

Comment:
This method will now look up the account information provided by the request and verify that the account exists in the account manager. if these checks fail then we know for certain the broker does not have the account in its cache so we can throw a ui required exception without sending any requests to the broker.

Reply:
Please rename the method and add a comment so that it is clear that it is only releavent for Silent requests. Non-silent request logic should not throw this exception.